### PR TITLE
Fix 4252 banner plugin placeholder

### PR DESCRIPTION
**Pull request by Travmatth**
**What kind of change does this PR introduce?**
adds feature to BannerPlugin, can now interpolate banner with the following values: [hash(:length)], [chunkhash(:length)], [name], and [ext] 

**Did you add tests for your changes?**
yes

**Summary**
Intends to resolve: https://github.com/webpack/webpack/issues/4252

**Does this PR introduce a breaking change?**
No

**Other information**
Change intended to allow interpolation of values listed above into the banner of a chunk. I would have added support for additional hash schemes (as is available in loader-utils.interpolateName function) but I am unsure as to which files are used to generate a hash.

### CODE REVIEW

**Comment by Travmatth** [comment]
Also, I think I've found some contradictory logic in beautify-lint - not sure if this is cause by rules defined inside the webpack repo or if these flaws are inside beautify-lint. If anyone can confirm they are inside this repo I can open another issue.

When returning an arrow function from inside a function, the following code lints with an error:
```
 const getReplacer = (value) => {
-       return (match) => `${value}`;
+       return(match) => `${value}`;
 };
```

But correcting this error reports another: 
```
$ME/webpack/lib/BannerPlugin.js
  35:2  error  Expected space(s) after "return"  keyword-spacing

✖ 1 problem (1 error, 0 warnings)
```

File: lib/BannerPlugin.js [hunk 1]
```diff
@@ -26,21 +69,20 @@ class BannerPlugin {
...
 					chunk.files
 						.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options))
-						.forEach((file) =>
-							compilation.assets[file] = new ConcatSource(
-								banner, "\n", compilation.assets[file]
-							)
-						);
+						.forEach((file) => {
+							const chunkHash = chunk.renderedHash || chunk.hash;
+							const comment = interpolate(banner, file, compilation.hash, chunkHash);
```
Comment (sokra) [inline lib/BannerPlugin.js]: please use the interpolation function which is also used by other stuff: `compilation.getPath`

File: lib/BannerPlugin.js [hunk 2]
```diff
@@ -8,10 +8,53 @@
...
 const ModuleFilenameHelpers = require("./ModuleFilenameHelpers");
 
-function wrapComment(str) {
+const wrapComment = (str) => {
 	if(!str.includes("\n")) return `/*! ${str} */`;
 	return `/*!\n * ${str.split("\n").join("\n * ")}\n */`;
-}
+};
+
+const REGEXP_HASH = /\[hash:?(\d+)?\]/gi,
```
Comment (sokra) [inline lib/BannerPlugin.js]: don't copy from other files

File: lib/BannerPlugin.js [hunk 3]
```diff
@@ -26,21 +26,36 @@ class BannerPlugin {
 	}
 
 	apply(compiler) {
-		const options = this.options;
-		const banner = this.banner;
+		const options = this.options,
+			banner = this.banner;
```
Comment (sokra) [inline lib/BannerPlugin.js]: please make separate declaration for each variable

File: test/configCases/plugins/banner-plugin-hashing/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+var source = require("fs")
+	.readFileSync(__filename, "utf-8")
+	.split("\n")
+	.slice(0,1)[0];
+
+const banner = parseBanner(source)
+
+it("should interpolate file hash in bundle0 chunk", () => {
+	banner["hash"].should.not.equal("[hash]");
```
Comment (sokra) [inline test/configCases/plugins/banner-plugin-hashing/index.js]: please make the tests more strict. This test would also pass if all placeholders are replaced with `""`.

File: lib/BannerPlugin.js [hunk 5]
```diff
@@ -26,21 +26,36 @@ class BannerPlugin {
...
-
 					chunk.files
 						.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options))
-						.forEach((file) =>
-							compilation.assets[file] = new ConcatSource(
-								banner, "\n", compilation.assets[file]
-							)
-						);
+						.forEach((file) => {
+							let filename, basename;
```
Comment (sokra) [inline lib/BannerPlugin.js]: `path/to/file.js`

filename: `path/to/file.js`
basename: `file.js`

File: lib/BannerPlugin.js [hunk 6]
```diff
@@ -26,21 +26,36 @@ class BannerPlugin {
...
+							} else {
+								filename = file;
+								basename = "";
+							}
+
+							const comment = compilation.getPath(banner, {
+								hash: compilation.hash,
+								chunk,
+								filename,
+								basename,
```
Comment (sokra) [inline lib/BannerPlugin.js]: also provide `query` as filename may contain a querystring

File: lib/BannerPlugin.js [hunk 7]
```diff
@@ -33,14 +33,36 @@ class BannerPlugin {
...
+						.forEach((file) => {
+							let basename;
+							let query = "";
+							let filename = file;
+							const hash = compilation.hash;
+							const querySplit = filename.indexOf("?");
+
+							if(querySplit >= 0) {
+								query = filename.substr(querySplit);
+								filename = filename.substr(0, querySplit);
```
Comment (sokra) [inline lib/BannerPlugin.js]: Could you add a test with query string?

File: lib/BannerPlugin.js [hunk 8]
```diff
@@ -33,14 +33,36 @@ class BannerPlugin {
...
+							}
+
+							if(filename.indexOf("/") < 0) {
+								basename = filename;
+							} else {
+								basename = filename.substr(filename.lastIndexOf("/") + 1);
```
Comment (sokra) [inline lib/BannerPlugin.js]: Could you add a test with basename != filename?

File: .gitignore [hunk 9]
```diff
@@ -10,3 +10,4 @@
 *.log
 .idea
 .vscode
+test/configCases/plugins/banner-plugin-hashing/dist
```
Comment (sokra) [inline .gitignore]: This shouldn't be needed. Tests should write to `test/js/...`

File: test/configCases/plugins/banner-plugin-hashing/webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+	node: {
+		__dirname: false,
+		__filename: false
+	},
+	entry: {
+		'dist/banner': ["./index.js"],
+		vendors: ["./vendors.js"]
+	},
+	output: {
+		//path: path.join(__dirname, "dist"),
```
Comment (sokra) [inline test/configCases/plugins/banner-plugin-hashing/webpack.config.js]: remove this comment

**Comment by sokra** [comment]
Thanks

**Comment by Travmatth** [comment]
my first open source contribution. thank you!

**Pull request merged**